parser/src/java/org/apache/hadoop/hive/ql/parse/IdentifiersParser.g

```diff
@@ -151,11 +182,46 @@ expressionsNotInParenthesis[boolean isStruct, boolean forceStruct]
     -> {$more.tree}
     ;
 
-expressionPart[CommonTree t, boolean isStruct]
+expressionPart[CommonTree firstExprTree, boolean isStruct]
     :
     (COMMA expression)+
-    -> {isStruct}? ^(TOK_FUNCTION Identifier["struct"] {$t} expression+)
-    -> {$t} expression+
+    -> {isStruct}? ^(TOK_FUNCTION Identifier["struct"] {$firstExprTree} expression+)
+    -> {$firstExprTree} expression+
+    ;
+
+// Parses comma separated list of expressions with optionally specified aliases and store the aliases for further usage.
+// <expression> [<alias>] [, <expression> [<alias>]]
+firstExpressionsWithAlias
+@init { initAliases(); }
+    :
+    first=expression colAlias=identifier? (COMMA expressionWithAlias)*
+    -> {colAlias != null}? ^(TOK_FUNCTION Identifier["named_struct"] { adaptor.create(Identifier, addAlias($colAlias.tree.getText())) } {$first.tree} expressionWithAlias*)
```

Comment:
Could we maybe identify this construct with a specific token? Would that help to identify later in the plan generation this case?

Reply:
I changed the rule `valuesTableConstructor` by adding an alternative: if there are no aliases defined for any of the values let's parse the values clause in the lagacy way which transforms to array of structs.
This seems to revert lots of the q.out files.

Reply:
I removed the `name_struct` approach and developed a different one using `struct` and a new toke to identify aliases. This required some additional functionality in the `TypeCheckProcFactory`